lucene/core/src/java/org/apache/lucene/search/knn/KnnCollectorManager.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.lucene.search.knn;
+
+import java.io.IOException;
+import org.apache.lucene.search.KnnCollector;
+import org.apache.lucene.util.BitSet;
+
+/**
+ * KnnCollectorManager responsible for creating {@link KnnCollector} instances. Useful to create
+ * {@link KnnCollector} instances that share global state across leaves, such a global queue of
+ * results collected so far.
+ */
+public abstract class KnnCollectorManager<C extends KnnCollector> {
+
+  /**
+   * Return a new {@link KnnCollector} instance.
+   *
+   * @param visitedLimit the maximum number of nodes that the search is allowed to visit
+   * @param parentBitSet the parent bitset, {@code null} if not applicable
+   */
+  public abstract C newCollector(int visitedLimit, BitSet parentBitSet) throws IOException;
```

Comment:
```suggestion
  public abstract C newCollector(int visitedLimit, LeafReaderContext context) throws IOException;
```

Also, I am not even sure `visitedLimit` should be there. It seems like something the manager should already know about (as in this instance its static) and we just need to know about the context (the context is for `DiversifyingChildrenFloatKnnVectorQuery` so that its collector manager can create `BitSet parentBitSet` from its encapsulated `BitSetProducer`).

I also think this method could return `null` if collection is not applicable for that given leaf context.

Reply:
`visitedLimit` changes per segment as different [cost](https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/search/AbstractKnnVectorQuery.java#L122)  of using filter in a segment context